### PR TITLE
Run translation validation on tracing error.

### PR DESCRIPTION
**Pull request by ysiraichi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106645

This PR wraps `InstructionTranslator` run with a try-catch block so as to run the
translation validation (TV) if it ends up raising an error.

In this context, we run TV so as to catch simplification errors. These may turn
`ShapeEnv.divisible` and `ShapeEnv.replacements` incorrect.

For example: #101173 describes a SymPy simplification bug that doesn't reach TV, since
it's run only in the end of the tracing.

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

**Comment by ysiraichi** [comment]
@ezyang Not sure how I should test this. Any ideas?

File: torch/_dynamo/convert_frame.py [hunk 1]
```diff
@@ -412,8 +412,19 @@ def transform(instructions, code_options):
...
+        try:
+            with tracing(tracer.output.tracing_context):
+                tracer.run()
+        except Exception:
+            if config.translation_validation:
+                fakes = tracer.output.tracked_fakes
+                tracer.output.shape_env.produce_guards(
+                    [a.fake for a in fakes],
+                    [a.source for a in fakes],
+                )
```
Comment (ezyang) [inline torch/_dynamo/convert_frame.py:425]: You might need to make source/target insertion transactional. What if an exception is raised between adding a source assert but before you added the target assert? Then TV would complain, even though there wasn't any problem.

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -3161,119 +3165,144 @@ def evaluate_expr(self, orig_expr: "sympy.Expr", hint=None, fx_node=None):
...
+        # no error will be raised until the end of this function.
+        #
+        # Reason: the translation validation may become invalid otherwise.
+        #
+        # If an error is raised before the end of this function, we remove the FX node
+        # inserted, and re-raise the error.
+        guard = None
+        tb = None
+
+        try:
```
Comment (ysiraichi) [inline torch/fx/experimental/symbolic_shapes.py:3193]: @ezyang I believe this try-catch block makes source/target guards transactional. Basically, if there's any exception raised before the end of this function, we remove the inserted source guard (by removing its corresponding `torch._assert` node).

File: torch/_dynamo/convert_frame.py (hunk 1 above)
Comment (ezyang) [reply torch/_dynamo/convert_frame.py:430]: You probably want to log the old exception here if a new exception is raised, so you don't "lose" the original error that triggered the TV.

**Review by ezyang — APPROVED** [review]

File: torch/_dynamo/convert_frame.py (hunk 1 above)
Comment (ysiraichi) [reply torch/_dynamo/convert_frame.py:436]: Yeah. I added `ValidationException` to the `except` block that does that.

**Pull request closed without merging**